File: src/engine/mahjong/tiles.ts
/**
 * 长沙麻将牌编码
 * 0-8 万, 9-17 条, 18-26 筒；每种 4 张 = 108 张（经典长沙不带字牌）
 */

export const SUIT_NAMES = ["万", "条", "筒"] as const;

export function tileName(t: number): string {
  if (t < 0 || t > 26) return `?${t}`;
  const suit = Math.floor(t / 9);
  const rank = (t % 9) + 1;
  return `${rank}${SUIT_NAMES[suit]}`;
}

export function buildDeck(): number[] {
  const deck: number[] = [];
  for (let t = 0; t < 27; t++) {
    for (let i = 0; i < 4; i++) deck.push(t);
  }
  return deck;
}

export function shuffle<T>(arr: T[], rng = Math.random): T[] {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

export function sortTiles(tiles: number[]): number[] {
  return [...tiles].sort((a, b) => a - b);
}

export function countsOf(tiles: number[]): number[] {
  const c = new Array(27).fill(0);
  for (const t of tiles) c[t]++;
  return c;
}

/** 标准 3N+2 胡牌（无癞子） */
export function canHu(tiles: number[]): boolean {
  if (tiles.length % 3 !== 2) return false;
  const c = countsOf(tiles);
  return canHuCounts(c);
}

function canHuCounts(c: number[]): boolean {
  for (let i = 0; i < 27; i++) {
    if (c[i] < 2) continue;
    c[i] -= 2;
    if (canMeldAll(c)) {
      c[i] += 2;
      return true;
    }
    c[i] += 2;
  }
  return false;
}

function canMeldAll(c: number[]): boolean {
  let i = 0;
  while (i < 27 && c[i] === 0) i++;
  if (i === 27) return true;
  // 刻子
  if (c[i] >= 3) {
    c[i] -= 3;
    if (canMeldAll(c)) {
      c[i] += 3;
      return true;
    }
    c[i] += 3;
  }
  // 顺子（同花色）
  const suit = Math.floor(i / 9);
  const rank = i % 9;
  if (rank <= 6 && c[i] > 0 && c[i + 1] > 0 && c[i + 2] > 0) {
    const base = suit * 9;
    if (i >= base && i + 2 < base + 9) {
      c[i]--;
      c[i + 1]--;
      c[i + 2]--;
      if (canMeldAll(c)) {
        c[i]++;
        c[i + 1]++;
        c[i + 2]++;
        return true;
      }
      c[i]++;
      c[i + 1]++;
      c[i + 2]++;
    }
  }
  return false;
}

/** 七对（长沙常见可选，本 Demo 开启） */
export function canQiDui(tiles: number[]): boolean {
  if (tiles.length !== 14) return false;
  const c = countsOf(tiles);
  let pairs = 0;
  for (let i = 0; i < 27; i++) {
    if (c[i] === 2) pairs++;
    else if (c[i] !== 0) return false;
  }
  return pairs === 7;
}

export function isHu(tiles: number[]): boolean {
  return canHu(tiles) || canQiDui(tiles);
}

/** 听牌提示：打哪张后仍可能听（简化：是否可胡某张） */
export function waitingTiles(hand: number[]): number[] {
  const waits: number[] = [];
  for (let t = 0; t < 27; t++) {
    if (isHu([...hand, t])) waits.push(t);
  }
  return waits;
}

File: server/lualib/game/changsha_mj/tiles.lua
-- 长沙麻将牌：0-8万 9-17条 18-26筒，各4张 = 108
local M = {}

function M.tile_name(t)
  if type(t) ~= "number" or t < 0 or t > 26 then return "?" .. tostring(t) end
  local suit = ({ "万", "条", "筒" })[math.floor(t / 9) + 1]
  return string.format("%d%s", (t % 9) + 1, suit)
end

function M.build_deck()
  local d = {}
  for t = 0, 26 do
    for _ = 1, 4 do d[#d + 1] = t end
  end
  return d
end

function M.shuffle(arr)
  local a = {}
  for i, v in ipairs(arr) do a[i] = v end
  for i = #a, 2, -1 do
    local j = math.random(i)
    a[i], a[j] = a[j], a[i]
  end
  return a
end

function M.sort_tiles(tiles)
  local a = {}
  for i, v in ipairs(tiles) do a[i] = v end
  table.sort(a)
  return a
end

function M.counts(tiles)
  local c = {}
  for i = 0, 26 do c[i] = 0 end
  for _, t in ipairs(tiles) do
    c[t] = (c[t] or 0) + 1
  end
  return c
end

local function can_meld_all(c)
  local i = 0
  while i <= 26 and (c[i] or 0) == 0 do i = i + 1 end
  if i > 26 then return true end
  local n = c[i]
  if n >= 3 then
    c[i] = n - 3
    if can_meld_all(c) then c[i] = n; return true end
    c[i] = n
  end
  local rank = i % 9
  if rank <= 6 and n > 0 and (c[i + 1] or 0) > 0 and (c[i + 2] or 0) > 0 then
    c[i] = n - 1
    c[i + 1] = c[i + 1] - 1
    c[i + 2] = c[i + 2] - 1
    if can_meld_all(c) then
      c[i] = n
      c[i + 1] = c[i + 1] + 1
      c[i + 2] = c[i + 2] + 1
      return true
    end
    c[i] = n
    c[i + 1] = c[i + 1] + 1
    c[i + 2] = c[i + 2] + 1
  end
  return false
end

--- 标准 3N+2 胡（无癞子）
function M.can_hu(tiles)
  if #tiles % 3 ~= 2 then return false end
  local c = M.counts(tiles)
  for eye = 0, 26 do
    if (c[eye] or 0) >= 2 then
      c[eye] = c[eye] - 2
      if can_meld_all(c) then
        c[eye] = c[eye] + 2
        return true
      end
      c[eye] = c[eye] + 2
    end
  end
  return false
end

--- 将将胡：全是 2/5/8
function M.is_jiang_tile(t)
  local r = (t % 9) + 1
  return r == 2 or r == 5 or r == 8
end

function M.can_jiang_jiang_hu(tiles)
  if #tiles % 3 ~= 2 then return false end
  for _, t in ipairs(tiles) do
    if not M.is_jiang_tile(t) then return false end
  end
  return M.can_hu(tiles)
end

--- 吃：仅上家；返回可选组合 {{t1,t2,discard}, ...} 手牌里要拿的两张
function M.chi_options(hand, discard)
  local opts = {}
  local c = M.counts(hand)
  local suit = math.floor(discard / 9)
  local rank = discard % 9 -- 0-8
  local base = suit * 9
  local patterns = {
    { rank - 2, rank - 1 },
    { rank - 1, rank + 1 },
    { rank + 1, rank + 2 },
  }
  for _, p in ipairs(patterns) do
    local a, b = p[1], p[2]
    if a >= 0 and b <= 8 then
      local t1, t2 = base + a, base + b
      if (c[t1] or 0) > 0 and (c[t2] or 0) > 0 then
        opts[#opts + 1] = { t1, t2, discard }
      end
    end
  end
  return opts
end

function M.remove_one(hand, tile)
  for i, t in ipairs(hand) do
    if t == tile then
      table.remove(hand, i)
      return true
    end
  end
  return false
end

function M.remove_n(hand, tile, n)
  for _ = 1, n do
    if not M.remove_one(hand, tile) then return false end
  end
  return true
end

return M
